src/libs/estimate/reestimate.ts

```diff
@@ -0,0 +1,31 @@
+export async function reestimate(fetchRequests: Function, counter: number = 0): Promise<any> {
+  // stop the execution on 5 fails;
+  // the below error message is not shown to the user so we are safe
+  if (counter >= 5)
+    return new Error(
+      'Estimation failure, retrying in a couple of seconds. If this issue persists, please change your RPC provider or contact Ambire support'
+    )
+
+  const estimationTimeout = new Promise((resolve) => {
+    setTimeout(() => {
+      resolve('Timeout reached')
+    }, 15000)
+  })
+
+  // try to estimate the request with a given timeout.
+  // if the request reaches the timeout, it cancels it and retries
+  let result = await Promise.race([Promise.all(fetchRequests()), estimationTimeout])
+
+  if (typeof result === 'string') {
+    const incremented = counter + 1
+    result = await reestimate(fetchRequests, incremented)
+  }
+
+  // if one of the calls returns an error, return it
+  if (Array.isArray(result)) {
```

Comment:
then if my suggestion is done this `Array.isArray` shouldn't be needed, it will just be `if (result === santinelTimeoutErr) ... else {`